pkg/reconciler/workspaceindex/index.go

```diff
@@ -0,0 +1,150 @@
+		klog.Infof("workspace %s/%s not scheduled, skipping...", workspace.ClusterName, workspace.Name)
+		return nil
+	}
+	history, err := convertHistory(workspace.Status.Location.History)
+	if err != nil {
+		return fmt.Errorf("invalid history for workspace %s/%s: %w", workspace.ClusterName, workspace.Name, err)
+	}
+	// TODO: how to handle the root itself?
+	org := workspace.ClusterName
+	if workspace.ClusterName != helper.OrganizationCluster {
+		orgName, workspaceName, err := helper.ParseLogicalClusterName(workspace.ClusterName)
+		if err != nil {
+			klog.Errorf("failed to determine org for cluster: %v", err)
+			return nil
+		}
+		if orgName != helper.OrganizationCluster {
+			klog.Errorf("invalid org %q, only one level of nesting allowed", orgName)
+			return nil
+		}
+		org = workspaceName
+	}
+
+	i.Lock()
+	if _, ok := i.workspaceMapping[org]; !ok {
+		i.workspaceMapping[org] = map[string][]ShardAssignment{}
+	}
+	i.workspaceMapping[org][workspace.Name] = history
+	klog.Infof("added history for %s->%s:%v", org, workspace.Name, history)
+	i.Unlock()
+	return nil
+}
+
+func convertHistory(original []tenancyv1alpha1.ShardStatus) ([]ShardAssignment, error) {
+	var history []ShardAssignment
+	for i := range original {
+		converted, err := convertShardStatus(fmt.Sprintf("history[%d]", i), original[i])
+		if err != nil {
+			return nil, err
+		}
+		history = append(history, converted)
+	}
+	return history, nil
+}
+
+func convertShardStatus(prefix string, original tenancyv1alpha1.ShardStatus) (ShardAssignment, error) {
+	var liveBeforeResourceVersion int64
+	if original.LiveBeforeResourceVersion == "" {
+		liveBeforeResourceVersion = math.MaxInt64
+	} else {
+		var err error
+		liveBeforeResourceVersion, err = strconv.ParseInt(original.LiveBeforeResourceVersion, 10, 64)
+		if err != nil {
+			return ShardAssignment{}, fmt.Errorf("%s.liveBeforeResourceVersion: invalid: %w", prefix, err)
+		}
+	}
+	liveAfterResourceVersion, err := strconv.ParseInt(original.LiveAfterResourceVersion, 10, 64)
+	if err != nil {
+		return ShardAssignment{}, fmt.Errorf("%s.liveAfterResourceVersion: invalid: %w", prefix, err)
+	}
+	return ShardAssignment{
+		Name:                      original.Name,
+		LiveBeforeResourceVersion: liveBeforeResourceVersion,
+		LiveAfterResourceVersion:  liveAfterResourceVersion,
+	}, nil
+}
+
+func (i *index) Get(organization, workspace string) ([]ShardAssignment, error) {
+	i.RLock()
+	defer i.RUnlock()
+	historyByWorkspace, organizationExists := i.workspaceMapping[organization]
+	if !organizationExists {
+		return nil, fmt.Errorf("organization %q not found", organization)
+	}
+	history, workspaceExists := historyByWorkspace[workspace]
+	if !workspaceExists {
+		return nil, fmt.Errorf("workspace %q not found", workspace)
+	}
+	return history, nil
+}
+
+func (i *index) List() map[string]map[string][]ShardAssignment {
+	i.RLock()
+	defer i.RUnlock()
+	return i.workspaceMapping
```

Comment:
Was I correct that this is only used to write out json?

Reply:
Yes!

Reply:
I'd say for now, even though this code exists for the benefit of e2e testing, we can still be somewhat defensive and implement json.Marshaler instead of handing out the map to callers. WDYT?

Reply:
> This gives back the underlying map which the caller could then destructively mutate

A go doc about the "this is not owned by the caller" would help.